### PR TITLE
O(1) time complexity subscribeKey and useSnapshot for big N keys proxy

### PR DESCRIPTION
**Pull request by pilaoda**
## Related Bug Reports or Discussions

#1160

## Summary

All tests pass, but I'm still not sure I handle all edge cases due to unfamiliar. Please help me check and add some test cases.

## Check List

- [ ] `pnpm run fix` for formatting and linting code and docs

### CODE REVIEW

File: tests/getter.test.tsx [hunk 1]
```diff
@@ -49,7 +49,7 @@ describe('getter', () => {
     await act(() => vi.advanceTimersByTimeAsync(0))
     expect(screen.getByText('A count: 2')).toBeInTheDocument()
     expect(screen.getByText('B count: 2')).toBeInTheDocument()
-    expect(computeDouble).toBeCalledTimes(1)
+    expect(computeDouble).toBeCalledTimes(4)
```
[the remark this answers is not in the document]
Comment (pilaoda) [reply tests/getter.test.tsx:52]: No, becasue we need to compute it on runtime, to track `count`.

[the remark this answers is not in the document]
Comment (dai-shi) [reply tests/getter.test.tsx:52]: Possible or impossible?

If you change tests, "All tests pass" is not true.

[the remark this answers is not in the document]
Comment (pilaoda) [reply tests/getter.test.tsx:52]: Impossible, you are right that all test pass is not true. This is a design desicion i made.

[the remark this answers is not in the document]
Comment (dai-shi) [reply tests/getter.test.tsx:52]: From the reactive system perspective, I think keeping this behavior is pretty important.
Your use case probably doesn't care about it, but that means we don't have a one-size-fits-all solution.

[the remark this answers is not in the document]
Comment (pilaoda) [reply tests/getter.test.tsx:52]: How this break the reactive rule? It just a sideeffect on render time, valtio already do side effect on render time in proxy getter.

[the remark this answers is not in the document]
Comment (pilaoda) [reply tests/getter.test.tsx:52]: In javascript, if users write a getter function, in their mind it is expected to compute on accessing in nature. 
Current pre compute and cache behavior in `createSnapshotDefault` some how let users confused no matter he is familiar with react or js only.

[the remark this answers is not in the document]
Comment (dai-shi) [reply tests/getter.test.tsx:52]: Isn't it a design decision? Or, are you saying `4` is better (less confusing) than `1`? What I learned is that computed values shouldn't be re-computed unless necessary. So, it's different from JS native getter behavior, which is a design decision.

[the remark this answers is not in the document]
Comment (pilaoda) [reply tests/getter.test.tsx:52]: It is a design decision for this O(1) `useSnapshot` to work. If we pre-compute and cache in `createSnapshotDefault`, all property the `getter` funciton used inside will disconnect the relation with getter property, but we need to track `count` dynamically if we access `doubled`.
Because we don't use `isChange` anymore, the change of `count` will not notify the subscribeKey on `doubled`.
If we want no `re-computed unless necessary` (sorry I didn't think about it), we can find some way to cache it after we compute it, but the first compute time still need to be on access explicitly.

[the remark this answers is not in the document]
Comment (pilaoda) [reply tests/getter.test.tsx:52]: So far if we want to cache the computed `doubled`, it should be hook scope level, because other component (other hook) create the snapshot may not need to track `count`. 
Then the called times still need change to `2` instead of keeping `1`.

[the remark this answers is not in the document]
Comment (pilaoda) [reply tests/getter.test.tsx:52]: If we really need to keep `1` call times, we may need to maintain a attribute dependency graph in proxy scope, it would be another big job, I suggest such big job should be anotther PR.

[the remark this answers is not in the document]
Comment (pilaoda) [reply tests/getter.test.tsx:52]: I also doubt that cache after compute can work, because we may need to re-subscribe each time we re-render.

**Comment by pilaoda** [comment]
> In case you missed it, I added some comments in the discussion thread.
> 
> * [Some idea about useSnapshot performance #1160 (reply in thread)](https://github.com/pmndrs/valtio/discussions/1160#discussioncomment-14147028)
> * [Some idea about useSnapshot performance #1160 (reply in thread)](https://github.com/pmndrs/valtio/discussions/1160#discussioncomment-14147520)

Thanks. I truly missed.

File: src/vanilla.ts [hunk 2]
```diff
@@ -56,11 +56,16 @@ export type Snapshot<T> = T extends { $$valtioSnapshot: infer S }
...
+type AddPropListener = (
+  prop: string | number | symbol,
+  listener: Listener,
+) => RemoveListener
 
 type ProxyState = readonly [
   target: object,
   ensureVersion: (nextCheckVersion?: number) => number,
   addListener: AddListener,
+  addPropListener: AddPropListener,
```
Comment (dai-shi) [inline src/vanilla.ts:68]: Why can't you use addListener and check prop value / version in your callback function? I believe it can still accomplish O(1)-useSnapshot. (My goal is to allow your third-party library to work without changing vanilla.ts without changing much.)

Comment (pilaoda) [reply src/vanilla.ts:68]: Yes. `addPropListener` logic can move outside vanilla.ts, but I found `createSnapshot` still need to modify, so I just left `addPropListener` there until I got full solution in my mind.

Comment (dai-shi) [reply src/vanilla.ts:68]: Do you modify createSnapshot in vanilla.ts?
I'm not sure what's blocking for your O(1)-useSnapshot without modifying vanilla.ts. I hope you find a solution. If not, you can try as much as possible, and I hope we can find something from there.

Comment (dai-shi) [reply src/vanilla.ts:68]: So, is it only about getter? Do you mean it works if we only use plain objects? Then, it sounds like a different requirement from O(1)-useSnapshot.

Comment (pilaoda) [reply src/vanilla.ts:68]: Most likely, yes. And some edge case about ProxySet and ProxyMap. I havn't give it a try. 
Also there are some dependencies on `proxy-compare` in `vanilla.ts`, I replace it with my self version, since I can't get the interval variables in `proxy-compare`.

Comment (dai-shi) [reply src/vanilla.ts:68]: At this point, I'm fairly confident that a third-party library can implement O(1)-useSnapshot, maybe without proxy-compare.

Yeah, I agree that it's not super cool that vanilla.ts depends on proxy-compare. I'll see how it will be a breaking change if we drop it.

**Comment by dai-shi** [comment]
> Yeah, I agree that it's not super cool that vanilla.ts depends on proxy-compare. I'll see how it will be a breaking change if we drop it.

Had a closer look at it, but unfortunately it isn't easy to remove it especially for ref in snapshots without sacrificing some performance. So, I'd keep it as it for now. We may consider it in the future version.

It's suboptimal, but those are basically no-op if you don't use proxy-compare. So, I guess it's technically possible to create a third-party library for O(1) useSnapshot. But, otherwise forking should only be an option.

We'll keep our design to depend on proxy-compare for now, but in the future we will probably consider O(1) subscription with dropping proxy-compare as I understand it would help some use cases.

Closing this but some of contributions here might be helpful then.
